internal/controller/instaslice_controller.go

```diff
@@ -114,55 +114,104 @@ func (r *InstasliceReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	// failed pods are not deleted by InstaSlice, finalizer is removed so that user can
 	// delete the pod and later InstaSlice will clean up the MIG slice.
 	if pod.Status.Phase == v1.PodFailed && controllerutil.ContainsFinalizer(pod, finalizerOrGateName) {
-
-		errRemovingFinalizer := controllerutil.RemoveFinalizer(pod, finalizerOrGateName)
-		if !errRemovingFinalizer {
-			log.FromContext(ctx).Info("finalizer not deleted for ", "pod", pod.Name)
+		allocationFound := false
+		for _, instaslice := range instasliceList.Items {
+			for podUuid, allocation := range instaslice.Spec.Allocations {
+				if podUuid == string(pod.UID) {
```

Comment:
Let's reverse the logic here and continue....

The block here can be a function; it might clean it up.